### DISCUSSION TITLE
Document how window positioning is handled

### DISCUSSION THREAD

**Discussion by DemiMarie** [Supreeeme/xwayland-satellite#68] [Q&A]
How does xwayland-satellite handle window management?  @asahilina mentioned on one of her streams that Xwayland applications under Sommelier had lots of window management bugs, presumably because absolute window positioning could not be forwarded to the host.  Asahi Linux will be using x112virtgpu to work around this, but security concerns mean that this is almost certainly a non-option for @QubesOS.

**Comment by Supreeeme** [comment]
The current positioning rules are:
- All toplevels are told they are located at the top left of their current monitor (xdg_output). So all Xwayland toplevels on a monitor at 0x0 are told they're at 0x0, all toplevels on a monitor at 2000x1080 are told they're at 2000x1080, etc.
    - In X11 all monitors seem to be in the same coordinate plane so without this offset windows on different monitors wouldn't receive cursor input correctly.
- Non override redirect windows are determined to be toplevels and they are repositioned as outlined above. This is fine as they usually expect the WM to reposition them.
- Override redirect window positioning is respected if the window is determined to be an xdg_popup. If there is no currently focused Xwayland window, the window is created as a toplevel. Otherwise, it's created as a popup for the window that was last hovered over, or last focused if such a window doesn't exist.

These rules seem to work well with most applications I've tested. There are still a few kinks though, such as #36 and override redirect windows that want to appear at the edge of the monitor use the wrong location (Steam notification popups do this).
